middleware/http/wasm/basic/basic.go

```diff
@@ -1,33 +1,58 @@
 package basic
 
 import (
+	"context"
 	"encoding/json"
 	"log"
 	"os"
 	"path/filepath"
+	"runtime"
+	"strconv"
 
+	"github.com/dapr/components-contrib/middleware"
+	"github.com/dapr/kit/logger"
 	"github.com/pkg/errors"
 	"github.com/tetratelabs/wazero"
-	"github.com/tetratelabs/wazero/wasi"
+	"github.com/tetratelabs/wazero/wasi_snapshot_preview1"
 	"github.com/valyala/fasthttp"
+)
 
-	"github.com/dapr/components-contrib/middleware"
-	"github.com/dapr/kit/logger"
+// These are undocumented, but exported functions. See tinygo-org/tinygo#2788
+const (
+	wasmFuncMalloc  = "malloc"
+	wasmFuncFree    = "free"
+	wasmFuncRewrite = "rewrite"
 )
 
 type middlewareMetadata struct {
 	Path    string `json:"path"`
 	Runtime string `json:"runtime"`
 }
 
-// Middleware is an wasm basic middleware.
+// Middleware is a wasm basic middleware.
 type Middleware struct {
-	logger logger.Logger
+	logger  logger.Logger
+	runtime wazero.Runtime
 }
 
 // NewMiddleware returns a new wasm basic middleware.
-func NewMiddleware(logger logger.Logger) *Middleware {
-	return &Middleware{logger: logger}
+func NewMiddleware(logger logger.Logger) (ret *Middleware) {
+	// Create a new WebAssembly Runtime.
+	r := wazero.NewRuntimeWithConfig(wazero.NewRuntimeConfig().
+		// Enable WebAssembly 2.0 support, which is required for TinyGo 0.24+.
+		WithWasmCore2())
+	ret = &Middleware{logger: logger, runtime: r}
+	runtime.SetFinalizer(ret, (*Middleware).Close)
+
+	ctx := context.Background()
+
+	// TinyGo needs WASI to implement functions such as panic.
+	if _, err := wasi_snapshot_preview1.Instantiate(ctx, ret.runtime); err != nil {
+		_ = ret.Close()
+		// An error is unexpected as Instantiate won't fail on a new runtime.
+		panic(errors.Wrap(err, "failed to instantiate WASI"))
+	}
+	return
```

Comment:
I prefer to return an error rather than panic, letting the dapr runtime handles the error. So I'd more like keep this func as just `return &Middleware{logger: logger}` and remove the init of the wazero to `GetHandler` func, which would also keep the same behavior with the other middleware components.

Reply:
So you are ok initializing a new runtime per middleware? That will add unnecessary overhead per call, but fine to keep things simple to start.

Reply:
I will do it a different way using namespaces. while the above will never error I can see it is distracting!